server/src/main/java/org/apache/druid/server/http/MetadataResource.java

```diff
@@ -158,7 +158,7 @@ public Response getDatabaseSegments(
     if (datasources != null && !datasources.isEmpty()) {
       druidDataSources = druidDataSources.stream()
                                          .filter(src -> datasources.contains(src.getName()))
-                                         .collect(Collectors.toSet());
+                                         .collect(Collectors.toList());
```

Comment:
what happens if we don't collect at all? can you just return a Stream and have jackson handle the lazy array materialization? I think you can avoid the odd output stream and json factory stuff if jackson supports such a thing.

Reply:
@surekhasaharan please add a comment explaining that getDataSources() returns unique data sources, so expensive collection to Set is not required.

Reply:
@drcrallen this collection happens only in case a non-null `datasources` query param is passed, else this part is skipped. This api already returns a stream, I think even if I remove the output stream and json factory stuff, this particular collection would remain, as it's used to filter out the datasources. The outputStream was added [here](https://github.com/apache/incubator-druid/pull/6094/commits/dc9fa4c40a5c725a1953269ab76ceac42cd91693#diff-eae30d4873d67410c81c288bcf7e8849R165), when I was writing the original code, and IIRC, I was getting some json format exceptions in `JsonParserIterator`.

Reply:
I don't think the collection is needed. The only place it is used is to immediately call `.stream()` on the next line. IMHO the more functional refactoring would be to produce  `Stream<DataSegment> metadataSegments` by optionally applying a filter rather than trying to manage the Collections. 

Not a blocker from my side, but I would like to see a conscious choice around it rather than letting it slide for a small patch

Reply:
@drcrallen fair point, I didn't see that it can be just filtered while creating `Stream<DataSegment> metadataSegments`, will update, then we don't need this collection. thanks.

Reply:
Also I got rid of outputstream and jsonfactory, tested it in local cluster and it seems to work, will keep it this way unless it breaks something.